### PR TITLE
Disk storage

### PR DESCRIPTION
**Pull request by NghiaTranUIT**
Disk Storage features

## Done
+ DiskPersistentHandler is responsible for various file-related tasks.
+ DiskStorage is responsible for Business Logic, such as Expiry, MaxTotal, Safe-Thread, ...
+ 100% test 

## Not yet
+ Expiry
+ Max Total

## Discuss
Please take a look a codebase. Honestly, I hate the syntax of the try-catch block. However, if not, we can't explicit what kind of error.
We should make up our mind.
How do you think?

### CODE REVIEW

**Comment by khoi** [comment]
Rebase to `master`, install `Swiftlint` and lint your code first.

**Comment by LarsJK** [comment]
I think [Metric](https://github.com/khoiln/swift-statsd-client/blob/master/Sources/Metrics/Metric.swift) should require serializable, and it needs to be implemented for all the metric types. Alternatively you could use Codable and get It for free.

When storing in the cache directory you could loose metrics because the system will delete when it needs to. Is that okay or should the documents directory be used?

File: Sources/Storage/Disk/DiskStorage.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        return queue.syncWithReturnedValue {
+            try? handler.get(key: key, type: Element.self)
+        }
+    }
+
+    func set(item: Element, forKey key: Key) {
+      queue.async(flags: .barrier) { [unowned self] in
+
+            // Don't handle Throws error here
+            try? self.handler.write(item, key: key, attribute: nil)
```
Comment (LarsJK) [inline Sources/Storage/Disk/DiskStorage.swift]: Would it be better to crash or handle the error? This will just silently fail to write to the cache.

Comment (NghiaTranUIT) [reply Sources/Storage/Disk/DiskStorage.swift]: yup, I agree with you. At that time I can't throw an exception from this async Block, so I just put it temporarily. Let me think an alternative approach to handle it

**Comment by NghiaTranUIT** [comment]
The reasons we implemented the Serializable is that we would like to support iOS 8.
Whereas, `Codable` is only available > iOS 9.0

**Comment by LarsJK** [comment]
I thought it would be available on any system as long as it is compiled with swift 4, since it is a language feature and not os feature.. I am probably wrong tho 😊

**edit:**
Does work in the simulator on 8.1

**Comment by khoi** [comment]
@LarsJK   @NghiaTranUIT 
>When storing in the cache directory you could loose metrics because the system will delete when it needs to. Is that okay or should the documents directory be used?

I'm still considering the pros and cons of this. In some cases, it's ok to lose some metrics since at the large scale, those numbers wont have any affect on the total analytics.

However, we'll try not to lose anything, and implement a `Flush` system, when the client try to flush everything out on Cocoa application cycles like `appDidEnterBackGround`.

That being said, let's have the flexibility of choosing either `.Documents` or `.Cache` directory, by accepting it as pass of the init param for DiskStorage

**Comment by khoi** [comment]
@NghiaTranUIT 
>The reasons we implemented the Serializable is that we would like to support iOS 8.
Whereas, Codable is only available > iOS 9.0

I also think that it's a Swift feature but I might be wrong. Where did you get the information saying it's only available from iOS 9 onward?

**Comment by khoi** [comment]
We'll be using Storage in https://github.com/khoiln/swift-statsd-client/issues/13

**Comment by NghiaTranUIT** [comment]
@khoiln @LarsJK You guys are right 👍. I'm totally wrong. It's a part of Standard of Swift library, not reply on specific OS version 🤣

**Comment by khoi** [comment]
Closed due to https://github.com/khoiln/swift-statsd-client/pull/15

**Pull request closed without merging**